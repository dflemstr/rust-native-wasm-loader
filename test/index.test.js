import webpack from 'webpack';
import path from 'path';
import nodeFileEval from 'node-file-eval';
import process from 'process';
import { execAsync } from 'async-child-process';
import { TextDecoder, TextEncoder } from 'text-encoding';

describe('rust-native-wasm-loader', () => {
  it('loads a simple cargo project', async () => {
    jest.setTimeout(600000);

    const options = {
      release: true,
    };

    const preRules = [{
      loader: 'wasm-loader'
    }];

    const stats = await runLoader('simple.js', 'simple', options, preRules);

    await expectToMatchSnapshot(stats);
  });

  it('loads a simple cargo project with warnings', async () => {
    jest.setTimeout(600000);

    const options = {
      release: true,
    };

    const preRules = [{
      loader: 'wasm-loader'
    }];

    // Run clean to ensure that the warning is generated every time
    await execAsync('cargo clean', {cwd: path.resolve(__dirname, 'fixtures', 'mywarninglib')});

    const stats = await runLoader('warning.js', 'warning', options, preRules);

    await expectToMatchSnapshot(stats);
  });

  it('loads a simple cargo project with errors', async () => {
    jest.setTimeout(600000);

    const options = {
      release: true,
    };

    const stats = await runLoader('error.js', 'error', options);

    await expectToMatchSnapshot(stats);
  });

  it('loads a simple cargo project with wasm-gc', async () => {
    jest.setTimeout(600000);

    const options = {
      release: true,
      gc: true,
    };

    const preRules = [{
      loader: 'wasm-loader'
    }];

    const stats = await runLoader('simple.js', 'simple-gc', options, preRules);

    await expectToMatchSnapshot(stats);
  });

  it('loads a cargo-web project', async () => {
    jest.setTimeout(600000);

    const options = {
      release: true,
      cargoWeb: {
        name: '[name].[hash:8].wasm',
      },
    };

    const stats = await runLoader('stdweb.js', 'stdweb', options);

    await expectToMatchSnapshot(stats);
  });

  it('loads a wasm-bindgen project', async () => {
    jest.setTimeout(600000);

    const options = {
      release: true,
      wasmBindgen: {
        wasm2es6js: true,
        nodejs: true,
      },
    };

    const stats = await runLoader('wasmbindgen.js', 'wasmbindgen', options);

    await expectToMatchSnapshot(stats);
  });

  it('loads a wasm-bindgen project with typescript support', async () => {
    jest.setTimeout(600000);

    const options = {
      release: true,
      wasmBindgen: {
        wasm2es6js: true,
        typescript: true,
        nodejs: true,
      },
    };

    const otherRules = [{
      test: /\.(js|rs|ts)$/,
      use: [{
        loader: 'ts-loader',
        options: {
          appendTsSuffixTo: [/\.rs$/],
          onlyCompileBundledFiles: true,
          colors: false
        }
      }]
    }];

    const stats = await runLoader('wasmbindgen.ts', 'wasmbindgen-ts', options, [], otherRules);

    await expectToMatchSnapshot(stats);
  });

  it('loads a wasm-bindgen project with typescript support and a type error', async () => {
    jest.setTimeout(600000);

    const options = {
      release: true,
      wasmBindgen: {
        wasm2es6js: true,
        typescript: true,
        nodejs: true,
      },
    };

    const otherRules = [{
      test: /\.(js|rs|ts)$/,
      use: [{
        loader: 'ts-loader',
        options: {
          appendTsSuffixTo: [/\.rs$/],
          onlyCompileBundledFiles: true,
          colors: false
        }
      }]
    }];

    const stats = await runLoader('wasmbindgen-type-error.ts', 'wasmbindgen-ts', options, [],
                                  otherRules);

    await expectToMatchSnapshot(stats);
  });

  it('loads a wasm-bindgen project with typescript support and a rust error', async () => {
    jest.setTimeout(600000);

    const options = {
      release: true,
      wasmBindgen: {
        wasm2es6js: true,
        typescript: true,
        nodejs: true,
      },
    };

    const otherRules = [{
      test: /\.(js|rs|ts)$/,
      use: [{
        loader: 'ts-loader',
        options: {
          appendTsSuffixTo: [/\.rs$/],
          onlyCompileBundledFiles: true,
          colors: false
        }
      }]
    }];

    const stats = await runLoader('wasmbindgen-rust-error.ts', 'wasmbindgen-ts', options, [],
                                  otherRules);

    await expectToMatchSnapshot(stats);
  });
});

function removeCWD(str) {
  return str.split(`${process.cwd()}/`).join('');
}

function cleanErrorStack(error) {
  return removeCWD(error.message.toString()).split('\n').filter(l => !l.match(/^\s+at\s/)).join('\n');
}

function collectDependencies(dependencies, obj) {
  if ('fileDependencies' in obj) {
    for (let dep of obj.fileDependencies) {
      dependencies.add(removeCWD(dep));
    }
  }

  if ('entries' in obj) {
    for (let entry of obj.entries) {
      collectDependencies(dependencies, entry);
    }
  }

  if ('module' in obj) {
    collectDependencies(dependencies, obj.module);
  }
}

async function expectToMatchSnapshot(stats) {
  const errors = stats.compilation.errors.map(cleanErrorStack);
  const warnings = stats.compilation.warnings.map(cleanErrorStack);
  const dependencies = new Set();

  collectDependencies(dependencies, stats.compilation);

  expect(errors).toMatchSnapshot('errors');
  expect(warnings).toMatchSnapshot('warnings');
  expect(dependencies).toMatchSnapshot('dependencies');

  const assetPath = stats.compilation.assets['index.js'].existsAt;
  try {
    const module = await nodeFileEval(assetPath, {
      encoding: 'utf-8',
      context: {
        require,
        Buffer,
        TextDecoder,
        TextEncoder,
        console,
        process,
        __dirname: path.dirname(assetPath)
      }
    });
    expect(await module.run()).toMatchSnapshot('output');
  } catch (e) {
    expect(e.message.split('\n')[0]).toMatchSnapshot('failure');
  }
}

function runLoader(fixture, test, options, preRules = [], otherRules = []) {
  const config = {
    context: path.resolve(__dirname, 'fixtures'),
    entry: `./${fixture}`,
    target: 'node',
    output: {
      path: path.resolve(__dirname, 'outputs', test),
      filename: 'index.js'
    },
    module: {
      rules: otherRules.concat([{
        test: /\.rs$/,
        use: preRules.concat([{
          loader: path.resolve(__dirname, '../src'),
          options,
        }])
      }])
    },
    node: {
      __dirname: false,
    }
  };

  const compiler = webpack(config);

  return new Promise((resolve, reject) => compiler.run((err, stats) => {
    if (err) {
      reject(err);
    } else {
      resolve(stats);
    }
  }))
}
