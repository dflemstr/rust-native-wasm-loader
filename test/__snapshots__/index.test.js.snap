// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rust-native-wasm-loader loads a cargo-web project: dependencies 1`] = `
Set {
  "test/fixtures/mystdweblib/Cargo.toml",
  "test/fixtures/mystdweblib/src/main.rs",
  "test/fixtures/stdweb.js",
}
`;

exports[`rust-native-wasm-loader loads a cargo-web project: errors 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a cargo-web project: output 1`] = `"Hello, World!"`;

exports[`rust-native-wasm-loader loads a cargo-web project: warnings 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a simple cargo project with errors: dependencies 1`] = `
Set {
  "test/fixtures/error.js",
  "test/fixtures/myerrorlib/Cargo.toml",
  "test/fixtures/myerrorlib/src/lib.rs",
}
`;

exports[`rust-native-wasm-loader loads a simple cargo project with errors: errors 1`] = `
Array [
  "Module build failed: BuildError: Error: Command failed: cargo build --message-format=json --target=wasm32-unknown-unknown --release
   Compiling myerrorlib v0.1.0 (test/fixtures/myerrorlib)
error: Could not compile \`myerrorlib\`.

To learn more, run the command again with --verbose.
",
  "error: aborting due to previous error

",
  "error[E0277]: cannot add \`&str\` to \`{integer}\`
 --> src/lib.rs:2:7
  |
2 |     1 + \\"two\\"
  |       ^ no implementation for \`{integer} + &str\`
  |
  = help: the trait \`std::ops::Add<&str>\` is not implemented for \`{integer}\`

",
]
`;

exports[`rust-native-wasm-loader loads a simple cargo project with errors: failure 1`] = `"Module build failed: BuildError: Error: Command failed: cargo build --message-format=json --target=wasm32-unknown-unknown --release"`;

exports[`rust-native-wasm-loader loads a simple cargo project with errors: warnings 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a simple cargo project with warnings: dependencies 1`] = `
Set {
  "test/fixtures/mywarninglib/Cargo.toml",
  "test/fixtures/mywarninglib/src/lib.rs",
  "test/fixtures/warning.js",
}
`;

exports[`rust-native-wasm-loader loads a simple cargo project with warnings: errors 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a simple cargo project with warnings: output 1`] = `1`;

exports[`rust-native-wasm-loader loads a simple cargo project with warnings: warnings 1`] = `
Array [
  "warning: unused variable: \`b\`
 --> src/lib.rs:2:20
  |
2 | pub fn add(a: i32, b: i32) -> i32 { // Warning: unused variable b
  |                    ^ help: consider using \`_b\` instead
  |
  = note: #[warn(unused_variables)] on by default

",
]
`;

exports[`rust-native-wasm-loader loads a simple cargo project with wasm-gc: dependencies 1`] = `
Set {
  "test/fixtures/myrustlib/Cargo.toml",
  "test/fixtures/myrustlib/src/add.rs",
  "test/fixtures/myrustlib/src/lib.rs",
  "test/fixtures/simple.js",
}
`;

exports[`rust-native-wasm-loader loads a simple cargo project with wasm-gc: errors 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a simple cargo project with wasm-gc: output 1`] = `3`;

exports[`rust-native-wasm-loader loads a simple cargo project with wasm-gc: warnings 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a simple cargo project: dependencies 1`] = `
Set {
  "test/fixtures/myrustlib/Cargo.toml",
  "test/fixtures/myrustlib/src/add.rs",
  "test/fixtures/myrustlib/src/lib.rs",
  "test/fixtures/simple.js",
}
`;

exports[`rust-native-wasm-loader loads a simple cargo project: errors 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a simple cargo project: output 1`] = `3`;

exports[`rust-native-wasm-loader loads a simple cargo project: warnings 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support and a rust error: dependencies 1`] = `
Set {
  "node_modules/@types/node/index.d.ts",
  "node_modules/@types/node/inspector.d.ts",
  "node_modules/typescript/lib/lib.d.ts",
  "test/fixtures/mywasmbindgenerrorlib/src/lib.rs",
  "test/fixtures/wasmbindgen-rust-error.ts",
}
`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support and a rust error: errors 1`] = `
Array [
  "[tsl] ERROR in test/fixtures/mywasmbindgenerrorlib/src/lib.rs.ts(1,17)
      TS2345: Argument of type '{ \\"killed\\": boolean; \\"code\\": number; \\"signal\\": null; \\"cmd\\": string; \\"stdout\\": string; \\"stderr\\": s...' is not assignable to parameter of type 'string'.",
  "[tsl] ERROR in test/fixtures/wasmbindgen-rust-error.ts(1,33)
      TS2306: File 'test/fixtures/mywasmbindgenerrorlib/src/lib.rs.ts' is not a module.",
  "Error: Command failed: cargo build --message-format=json --target=wasm32-unknown-unknown --release
   Compiling mywasmbindgenerrorlib v0.1.0 (test/fixtures/mywasmbindgenerrorlib)
error: Could not compile \`mywasmbindgenerrorlib\`.

To learn more, run the command again with --verbose.
",
  "error: aborting due to previous error

",
  "error[E0369]: binary operation \`+\` cannot be applied to type \`&str\`
help: \`to_owned()\` can be used to create an owned \`String\` from a string reference. String concatenation appends the string on the right to the string on the left and may require reallocation. This requires ownership of the string on the left
  |
1 | .to_owned()#![feature(proc_macro, wasm_custom_section, wasm_import_module)]
  | ^^^^^^^^^^^

",
]
`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support and a rust error: failure 1`] = `"[object Object]"`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support and a rust error: warnings 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support and a type error: dependencies 1`] = `
Set {
  "node_modules/@types/node/index.d.ts",
  "node_modules/@types/node/inspector.d.ts",
  "node_modules/typescript/lib/lib.d.ts",
  "test/fixtures/mywasmbindgenlib/src/lib.rs",
  "test/fixtures/wasmbindgen-type-error.ts",
}
`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support and a type error: errors 1`] = `
Array [
  "[tsl] ERROR in test/fixtures/mywasmbindgenlib/src/lib.rs.ts(1,17)
      TS2345: Argument of type '{ \\"killed\\": boolean; \\"code\\": number; \\"signal\\": null; \\"cmd\\": string; \\"stdout\\": string; \\"stderr\\": s...' is not assignable to parameter of type 'string'.",
  "[tsl] ERROR in test/fixtures/wasmbindgen-type-error.ts(1,33)
      TS2306: File 'test/fixtures/mywasmbindgenlib/src/lib.rs.ts' is not a module.",
  "Error: Command failed: cargo build --message-format=json --target=wasm32-unknown-unknown --release
   Compiling mywasmbindgenlib v0.1.0 (test/fixtures/mywasmbindgenlib)
error: Could not compile \`mywasmbindgenlib\`.

To learn more, run the command again with --verbose.
",
  "error: aborting due to 2 previous errors

",
  "error[E0635]: unknown feature \`wasm_custom_section\`
 --> src/lib.rs:1:24
  |
1 | #![feature(proc_macro, wasm_custom_section, wasm_import_module)]
  |                        ^^^^^^^^^^^^^^^^^^^

",
  "error[E0635]: unknown feature \`wasm_import_module\`
 --> src/lib.rs:1:45
  |
1 | #![feature(proc_macro, wasm_custom_section, wasm_import_module)]
  |                                             ^^^^^^^^^^^^^^^^^^

",
]
`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support and a type error: failure 1`] = `"[object Object]"`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support and a type error: warnings 1`] = `
Array [
  "warning: the feature \`proc_macro\` has been stable since 1.29.0 and no longer requires an attribute to enable
 --> src/lib.rs:1:12
  |
1 | #![feature(proc_macro, wasm_custom_section, wasm_import_module)]
  |            ^^^^^^^^^^
  |
  = note: #[warn(stable_features)] on by default

",
]
`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support: dependencies 1`] = `
Set {
  "node_modules/@types/node/index.d.ts",
  "node_modules/@types/node/inspector.d.ts",
  "node_modules/typescript/lib/lib.d.ts",
  "test/fixtures/mywasmbindgenlib/src/lib.rs",
  "test/fixtures/wasmbindgen.ts",
}
`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support: errors 1`] = `
Array [
  "[tsl] ERROR in test/fixtures/mywasmbindgenlib/src/lib.rs.ts(1,17)
      TS2345: Argument of type '{ \\"killed\\": boolean; \\"code\\": number; \\"signal\\": null; \\"cmd\\": string; \\"stdout\\": string; \\"stderr\\": s...' is not assignable to parameter of type 'string'.",
  "[tsl] ERROR in test/fixtures/wasmbindgen.ts(1,33)
      TS2306: File 'test/fixtures/mywasmbindgenlib/src/lib.rs.ts' is not a module.",
  "Error: Command failed: cargo build --message-format=json --target=wasm32-unknown-unknown --release
   Compiling mywasmbindgenlib v0.1.0 (test/fixtures/mywasmbindgenlib)
error: Could not compile \`mywasmbindgenlib\`.

To learn more, run the command again with --verbose.
",
  "error: aborting due to 2 previous errors

",
  "error[E0635]: unknown feature \`wasm_custom_section\`
 --> src/lib.rs:1:24
  |
1 | #![feature(proc_macro, wasm_custom_section, wasm_import_module)]
  |                        ^^^^^^^^^^^^^^^^^^^

",
  "error[E0635]: unknown feature \`wasm_import_module\`
 --> src/lib.rs:1:45
  |
1 | #![feature(proc_macro, wasm_custom_section, wasm_import_module)]
  |                                             ^^^^^^^^^^^^^^^^^^

",
]
`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support: failure 1`] = `"[object Object]"`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project with typescript support: warnings 1`] = `
Array [
  "warning: the feature \`proc_macro\` has been stable since 1.29.0 and no longer requires an attribute to enable
 --> src/lib.rs:1:12
  |
1 | #![feature(proc_macro, wasm_custom_section, wasm_import_module)]
  |            ^^^^^^^^^^
  |
  = note: #[warn(stable_features)] on by default

",
]
`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project: dependencies 1`] = `
Set {
  "test/fixtures/mywasmbindgenlib/Cargo.toml",
  "test/fixtures/mywasmbindgenlib/src/lib.rs",
  "test/fixtures/wasmbindgen.js",
}
`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project: errors 1`] = `
Array [
  "Error: Command failed: cargo build --message-format=json --target=wasm32-unknown-unknown --release
   Compiling mywasmbindgenlib v0.1.0 (test/fixtures/mywasmbindgenlib)
error: Could not compile \`mywasmbindgenlib\`.

To learn more, run the command again with --verbose.
",
  "error: aborting due to 2 previous errors

",
  "error[E0635]: unknown feature \`wasm_custom_section\`
 --> src/lib.rs:1:24
  |
1 | #![feature(proc_macro, wasm_custom_section, wasm_import_module)]
  |                        ^^^^^^^^^^^^^^^^^^^

",
  "error[E0635]: unknown feature \`wasm_import_module\`
 --> src/lib.rs:1:45
  |
1 | #![feature(proc_macro, wasm_custom_section, wasm_import_module)]
  |                                             ^^^^^^^^^^^^^^^^^^

",
]
`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project: failure 1`] = `"[object Object]"`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project: warnings 1`] = `
Array [
  "warning: the feature \`proc_macro\` has been stable since 1.29.0 and no longer requires an attribute to enable
 --> src/lib.rs:1:12
  |
1 | #![feature(proc_macro, wasm_custom_section, wasm_import_module)]
  |            ^^^^^^^^^^
  |
  = note: #[warn(stable_features)] on by default

",
]
`;
